### PR TITLE
Add new Hankel1 and BesselJ functions that work with CSE

### PR DESCRIPTION
**Pull request by isuruf**
Since Derivatives and Subs are not CSEed due to some issues
derivatives of hankel1 functions and hankel1 functions
end up with different arguments and code generation treats
them seperately. See https://github.com/inducer/sumpy/blob/0fc6535b6ef803e2a8ebd418162274634aead6f9/sumpy/cse.py#L84

Fixes https://github.com/inducer/sumpy/issues/61

### CODE REVIEW

File: sumpy/symbolic.py [hunk 1]
```diff
@@ -241,4 +241,53 @@ def map_subscript(self, expr):
...
+    arguments are (order, z, nderivs)"""
+    nargs = (3,)
+
+    def fdiff(self, argindex=1):
+        if argindex in (1, 3):
+            # we are not differentiating w.r.t order or nderivs
+            return 0
+        args = list(self.args)
+        args[-1] += 1
+        return self.func(*args)
```
Comment (inducer) [inline sumpy/symbolic.py:272]:
```suggestion
        order, z, nderivs = list(self.args)
        return self.func(order, z, nderivs+1)
```

File: sumpy/symbolic.py [hunk 2]
```diff
@@ -241,4 +241,53 @@ def map_subscript(self, expr):
...
+class _BesselOrHankel(sympy.Function):
+    """A symbolic function for BesselJ or Hankel1 functions
+    that keeps track of the derivatives taken of the function
+    arguments are (order, z, nderivs)"""
+    nargs = (3,)
+
+    def fdiff(self, argindex=1):
+        if argindex in (1, 3):
+            # we are not differentiating w.r.t order or nderivs
+            return 0
```
Comment (inducer) [inline sumpy/symbolic.py:269]:
```suggestion
            raise ValueError()
```
?
(Is this code path ever hit? I feel much better about errors than silent garbage data.)

**Review by inducer — COMMENTED** [review]
Thanks for investigating! Two questions/suggestions, otherwise LGTM!

File: sumpy/symbolic.py [hunk 3]
```diff
@@ -241,4 +241,53 @@ def map_subscript(self, expr):
...
+
+
+import sympy
+
+
+class _BesselOrHankel(sympy.Function):
+    """A symbolic function for BesselJ or Hankel1 functions
+    that keeps track of the derivatives taken of the function
+    arguments are (order, z, nderivs)"""
+    nargs = (3,)
```
Comment (inducer) [inline sumpy/symbolic.py:264]:
```suggestion
    that keeps track of the derivatives taken of the function.
    Arguments are ``(order, z, nderivs)``.
    """
    nargs = (3,)
```

File: sumpy/symbolic.py (hunk 2 above)
Comment (isuruf) [reply sumpy/symbolic.py:270]: Not sure. Let's try

**Comment by inducer** [comment]
Thanks!

**Pull request merged**